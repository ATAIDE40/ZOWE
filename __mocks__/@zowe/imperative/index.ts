/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

/**
 * Mocks all imperative classes specified under __mocks__/__imperative__
 * @returns mocked version of the @zowe/imperative module
 */
const __imperative__get = function () {
  const fs = require.requireActual("fs");
  const imperativePath = require("path").resolve(__dirname, "..", "..", "__imperative__");
  const extraClasses: any = {};
  fs.readdirSync(imperativePath).forEach((file: string) => {
    extraClasses[file.split(".")[0]] = require(`${imperativePath}/${file}`);
  });
  return {
    ...require.requireActual("@zowe/imperative"),
    // Note: Comment out the line below if you do not want to mock
    //       the classes in __mocks__/__imperative__/ for local testing purposes
    ...extraClasses
    // Note: You can still use Imperative.console.info("debug") in your test
  };
}

module.exports = __imperative__get();
