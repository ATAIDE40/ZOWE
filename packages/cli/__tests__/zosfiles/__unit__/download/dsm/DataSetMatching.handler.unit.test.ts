/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Download, IDownloadOptions } from "@zowe/zos-files-for-zowe-sdk";
import { UNIT_TEST_ZOSMF_PROF_OPTS } from "../../../../../../../__tests__/__src__/mocks/ZosmfProfileMock";

const fakeDownloadOptions: IDownloadOptions = {
    binary: undefined,
    directory: undefined,
    encoding: undefined,
    excludePatterns: undefined,
    extension: undefined,
    extensionMap: null,
    failFast: undefined,
    maxConcurrentRequests: undefined,
    preserveOriginalLetterCase: undefined,
    record: undefined,
    responseTimeout: undefined,
    volume: undefined
};

describe("Download DataSetMatching handler", () => {
    describe("process method", () => {
        it("should download matching datasets if requested", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../../../src/zosfiles/download/dsm/DataSetMatching.handler");
            const handler = new handlerReq.default();
            const pattern = "testing";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession = null;

            // Mock the submit JCL function
            Download.dataSetsMatchingPattern = jest.fn((session) => {
                fakeSession = session;
                return {
                    success: true,
                    commandResponse: "downloaded"
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.process({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        pattern,
                        ...UNIT_TEST_ZOSMF_PROF_OPTS
                    },
                    response: {
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        },
                        progress: {
                            startBar: jest.fn((parms) => {
                                // do nothing
                            }),
                            endBar: jest.fn(() => {
                                // do nothing
                            })
                        }
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeUndefined();
            expect(Download.dataSetsMatchingPattern).toHaveBeenCalledTimes(1);
            expect(Download.dataSetsMatchingPattern).toHaveBeenCalledWith(fakeSession, [pattern], {
                ...fakeDownloadOptions,
                task: {
                    percentComplete: 0,
                    stageName: 0,
                    statusMessage: "Downloading data sets matching a pattern"
                }
            });
            expect(jsonObj).toMatchSnapshot();
            expect(apiMessage).toMatchSnapshot();
            expect(logMessage).toMatchSnapshot();
        });

        it("should handle generation of an extension map", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../../../src/zosfiles/download/dsm/DataSetMatching.handler");
            const handler = new handlerReq.default();
            const pattern = "testing";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession = null;
            const extensionMap = "CNTL=JCL,PARMLIB=JCL,LOADLIB=JCL";

            // Mock the submit JCL function
            Download.dataSetsMatchingPattern = jest.fn((session) => {
                fakeSession = session;
                return {
                    success: true,
                    commandResponse: "downloaded"
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.process({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        pattern,
                        extensionMap,
                        ...UNIT_TEST_ZOSMF_PROF_OPTS
                    },
                    response: {
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        },
                        progress: {
                            startBar: jest.fn((parms) => {
                                // do nothing
                            }),
                            endBar: jest.fn(() => {
                                // do nothing
                            })
                        }
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeUndefined();
            expect(Download.dataSetsMatchingPattern).toHaveBeenCalledTimes(1);
            expect(Download.dataSetsMatchingPattern).toHaveBeenCalledWith(fakeSession, [pattern], {
                ...fakeDownloadOptions,
                extensionMap: {cntl: "jcl", parmlib: "jcl", loadlib: "jcl"},
                task: {
                    percentComplete: 0,
                    stageName: 0,
                    statusMessage: "Downloading data sets matching a pattern"
                }
            });
            expect(jsonObj).toMatchSnapshot();
            expect(apiMessage).toMatchSnapshot();
            expect(logMessage).toMatchSnapshot();
        });

        it("should gracefully handle an extension map parsing error", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../../../src/zosfiles/download/dsm/DataSetMatching.handler");
            const handler = new handlerReq.default();
            const pattern = "testing";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession = null;
            const extensionMap = "CNTL=JCL,PARMLIB-JCL,LOADLIB=JCL";

            // Mock the submit JCL function
            Download.dataSetsMatchingPattern = jest.fn((session) => {
                fakeSession = session;
                return {
                    success: true,
                    commandResponse: "downloaded"
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.process({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        pattern,
                        extensionMap,
                        ...UNIT_TEST_ZOSMF_PROF_OPTS
                    },
                    response: {
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        },
                        progress: {
                            startBar: jest.fn((parms) => {
                                // do nothing
                            }),
                            endBar: jest.fn(() => {
                                // do nothing
                            })
                        }
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeDefined();
            expect(error.message).toContain("An error occurred processing the extension map");
            expect(Download.dataSetsMatchingPattern).toHaveBeenCalledTimes(0);
            expect(jsonObj).toMatchSnapshot();
            expect(apiMessage).toMatchSnapshot();
            expect(logMessage).toMatchSnapshot();
        });

        it("should handle generation of an exclusion list", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../../../src/zosfiles/download/dsm/DataSetMatching.handler");
            const handler = new handlerReq.default();
            const pattern = "testing";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession = null;
            const excludePatterns = "TEST.EXCLUDE.**.CNTL";

            // Mock the submit JCL function
            Download.dataSetsMatchingPattern = jest.fn((session) => {
                fakeSession = session;
                return {
                    success: true,
                    commandResponse: "downloaded"
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.process({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        pattern,
                        excludePatterns,
                        ...UNIT_TEST_ZOSMF_PROF_OPTS
                    },
                    response: {
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        },
                        progress: {
                            startBar: jest.fn((parms) => {
                                // do nothing
                            }),
                            endBar: jest.fn(() => {
                                // do nothing
                            })
                        }
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeUndefined();
            expect(Download.dataSetsMatchingPattern).toHaveBeenCalledTimes(1);
            expect(Download.dataSetsMatchingPattern).toHaveBeenCalledWith(fakeSession, [pattern], {
                ...fakeDownloadOptions,
                excludePatterns: ["TEST.EXCLUDE.**.CNTL"],
                task: {
                    percentComplete: 0,
                    stageName: 0,
                    statusMessage: "Downloading data sets matching a pattern"
                }
            });
            expect(jsonObj).toMatchSnapshot();
            expect(apiMessage).toMatchSnapshot();
            expect(logMessage).toMatchSnapshot();
        });
    });
});
