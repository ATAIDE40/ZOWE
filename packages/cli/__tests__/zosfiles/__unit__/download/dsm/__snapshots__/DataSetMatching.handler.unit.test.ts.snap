// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download DataSetMatching handler process method should download matching datasets if requested 1`] = `
Object {
  "commandResponse": "downloaded",
  "success": true,
}
`;

exports[`Download DataSetMatching handler process method should download matching datasets if requested 2`] = `""`;

exports[`Download DataSetMatching handler process method should download matching datasets if requested 3`] = `
"
downloaded"
`;
